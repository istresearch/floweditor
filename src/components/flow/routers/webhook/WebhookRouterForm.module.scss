@import 'variables.module.scss';

.link {
  text-decoration: underline;
  color: $link_color;
  cursor: pointer;
}

.info {
  margin-bottom: 0px;
}

.method {
  margin-right: 10px;
  min-width: 110px;
}

.url {
  flex-grow: 1;
  input {
    height: 36px;
  }
}

.code {
  background: #f3f3f3;
  padding: 10px;
  -webkit-border-radius: 4px;
  -moz-border-radius: 4px;
  border-radius: 4px;
  font-size: 12px;
  font-family: 'Roboto Mono', monospace;
  font-weight: 300;
}

.instructions {
  margin-top: 10px;
}

.headers_title {
  margin-top: 0;
}

.remove_ico {
  color: #ccc;
  border: 0px solid green;
  opacity: 0;
  font-size: 12px;
  line-height: 12px;
  width: 20px;
  position: absolute;
  right: -23px;
  top: 25px;
}

.example {
  unicode-bidi: embed;
  font-size: 12px;
  font-family: 'Roboto Mono', monospace;
  font-weight: 300;
  white-space: pre;
  padding: 4px;
  font-size: 0.9em;
  background: $light_gray_2;
  -webkit-border-radius: 4px;
  -moz-border-radius: 4px;
  border-radius: 4px;
}

.header {
  input {
    height: 28px;
    line-height: 1.2em;
    font-size: 12px;
    padding: 2px 10px;
    border-color: #ccc;
  }
  padding-top: 8px;
  &:hover {
    .remove_ico {
      cursor: pointer;
      opacity: 1;
    }
  }
}

.header_name {
  width: 170px;
  display: inline-block;
}

.header_value {
  width: 434px;
  display: inline-block;
  margin-left: 10px;
}

.body_form {
  .req_body {
    margin-top: 8px;
    textarea {
      display: inline-block;
      padding: 5px 5px;
      unicode-bidi: embed;
      font-size: 12px !important;
      font-family: 'Roboto Mono', monospace;
      font-weight: 300;
      white-space: pre;
      min-height: 200px;
    }
  }
}
