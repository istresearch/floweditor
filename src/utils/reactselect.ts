import { StylesConfig } from 'react-select';

export const large = {
  placeholder: (styles: StylesConfig, state: any) => {
    return { ...styles, color: '#ddd', marginLeft: '3px' };
  },
  singleValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '3px' };
  },
  multiValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '0px' };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none',
      fontSize: '13px'
    };
  },
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      fontSize: '13px',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#e6e6e6'
      },
      '&:focus-within': {
        boxShadow: 'var(--widget-box-shadow-focused) !important',
        border: '1px solid #a4cafe !important'
      }
    };
  }
};

export const messageStyle = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      borderBottom: 'none',
      borderBottomRightRadius: '0px',
      borderBottomLeftRadius: '0px',
      '&:hover': {
        borderColor: '#e6e6e6'
      }
    };
  }
};

export const getErroredSelect = (baseControl: any): any => {
  return {
    control: (styles: StylesConfig) => ({
      ...styles,
      ...baseControl,
      borderColor: 'tomato',
      boxShadow: '0 0 0 3px rgba(255,196,186,0.5) !important;',
      '&:hover': {
        borderColor: 'tomato'
      }
    })
  };
};

export const small = {
  control: (styles: StylesConfig) => ({
    ...styles,
    borderColor: '#e6e6e6',
    boxShadow: 'none',
    '&:hover': {
      borderColor: '#e6e6e6'
    },
    maxHeight: '28px',
    minHeight: '28px',
    fontSize: '12px',
    lineHeight: '12px',
    marginBottom: '-10px',
    '&:focus-within': {
      boxShadow: 'var(--widget-box-shadow-focused) !important',
      border: '1px solid #a4cafe !important'
    }
  }),
  menuList: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 200
  }),
  option: (styles: StylesConfig) => ({
    ...styles,
    fontSize: 12
  }),
  singleValue: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 28,
    lineHeight: '13px'
  }),
  indicatorsContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  valueContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  input: (styles: StylesConfig) => ({
    ...styles,
    marginLeft: '-2px',
    marginTop: '0px',
    caretColor: '#999',
    boxShadow: 'none',
    border: 'none',
    maxHeight: '20px',
    minHeight: '20px',
    lineHeight: '12px'
  })
};

export const tagging = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#e6e6e6'
      },
      '&:focus-within': {
        boxShadow: 'var(--widget-box-shadow-focused) !important',
        border: '1px solid #a4cafe !important'
      }
    };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none'
    };
  },
  multiValue: (styles: StylesConfig) => {
    return {
      ...styles,
      maxHeight: 24,
      minHeight: 24,
      lineHeight: '17px',
      backgroundColor: 'hsl(0, 0%, 94%)',
      border: '1px solid hsl(0, 0%, 85%)'
    };
  }
};
