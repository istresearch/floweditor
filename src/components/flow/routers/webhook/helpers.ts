import { createWebhookBasedNode } from 'components/flow/routers/helpers';
import { WebhookRouterFormState } from 'components/flow/routers/webhook/WebhookRouterForm';
import { DEFAULT_BODY } from 'components/nodeeditor/constants';
import { Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { CallWebhook } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { createUUID } from 'utils';

export enum Methods {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  HEAD = 'HEAD',
  PATCH = 'PATCH'
}

export interface MethodOption {
  value: string;
  name: string;
}

interface HeaderMap {
  [name: string]: string;
}

export const GET_METHOD: MethodOption = {
  value: Methods.GET,
  name: Methods.GET
};

export const METHOD_OPTIONS: MethodOption[] = [
  GET_METHOD,
  { value: Methods.POST, name: Methods.POST },
  { value: Methods.PUT, name: Methods.PUT },
  { value: Methods.DELETE, name: Methods.DELETE },
  { value: Methods.HEAD, name: Methods.HEAD },
  { value: Methods.PATCH, name: Methods.PATCH }
];

export const getOriginalAction = (settings: NodeEditorSettings): CallWebhook => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 && settings.originalNode.node.actions[0]);

  if (action.type === Types.call_webhook) {
    return action as CallWebhook;
  }
};

export const nodeToState = (settings: NodeEditorSettings): WebhookRouterFormState => {
  // TODO: work out an incremental result name
  const resultName: StringEntry = { value: 'Result' };

  const state: WebhookRouterFormState = {
    headers: [],
    resultName,
    method: { value: GET_METHOD },
    url: { value: '' },
    body: { value: getDefaultBody(Methods.GET) },
    valid: false
  };

  if (getType(settings.originalNode) === Types.split_by_webhook) {
    const action = getOriginalAction(settings) as CallWebhook;

    // add in our headers
    for (const name of Object.keys(action.headers || []).sort()) {
      state.headers.push({
        value: {
          uuid: createUUID(),
          value: action.headers[name],
          name
        }
      });
    }

    state.resultName = { value: action.result_name };
    state.url = { value: action.url };
    state.method = { value: { name: action.method, value: action.method } };
    state.body = { value: action.body };
    state.valid = true;
  } else {
    state.headers.push({
      value: {
        uuid: createUUID(),
        name: 'Accept',
        value: 'application/json'
      }
    });
  }

  // one empty header
  state.headers.push({
    value: {
      uuid: createUUID(),
      name: '',
      value: ''
    }
  });

  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: WebhookRouterFormState
): RenderNode => {
  const headers: HeaderMap = {};

  for (const entry of state.headers) {
    if (entry.value.name.trim().length !== 0) {
      headers[entry.value.name] = entry.value.value;
    }
  }

  let uuid = createUUID();

  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const newAction: CallWebhook = {
    uuid,
    headers,
    type: Types.call_webhook,
    url: state.url.value,
    body: state.body.value,
    method: state.method.value.value as Methods,
    result_name: state.resultName.value
  };

  return createWebhookBasedNode(newAction, settings.originalNode, false);
};

export const getDefaultBody = (method: string): string => {
  return method === Methods.GET ? '' : DEFAULT_BODY;
};
