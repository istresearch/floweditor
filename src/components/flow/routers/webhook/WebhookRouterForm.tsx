import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet, Tab } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import HeaderElement, { Header } from 'components/flow/routers/webhook/header/HeaderElement';
import {
  METHOD_OPTIONS,
  MethodOption,
  Methods,
  nodeToState,
  stateToNode,
  getDefaultBody
} from 'components/flow/routers/webhook/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import SelectElement from 'components/form/select/SelectElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import * as React from 'react';
import { FormEntry, FormState, mergeForm, StringEntry, ValidationFailure } from 'store/nodeEditor';
import {
  Alphanumeric,
  shouldRequireIf,
  StartIsNonNumeric,
  validate,
  ValidURL,
  validateIf
} from 'store/validators';
import { createUUID } from 'utils';

import styles from './WebhookRouterForm.module.scss';
import { Trans } from 'react-i18next';
import i18n from 'config/i18n';

export interface HeaderEntry extends FormEntry {
  value: Header;
}

export interface MethodEntry extends FormEntry {
  value: MethodOption;
}

export interface WebhookRouterFormState extends FormState {
  headers: HeaderEntry[];
  method: MethodEntry;
  url: StringEntry;
  body: StringEntry;
  resultName: StringEntry;
}

export default class WebhookRouterForm extends React.Component<
  RouterFormProps,
  WebhookRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);
    this.state = nodeToState(this.props.nodeSettings);
    bindCallbacks(this, {
      include: [/^handle/]
    });
  }

  private handleUpdate(
    keys: {
      method?: MethodOption;
      url?: string;
      body?: string;
      header?: Header;
      removeHeader?: Header;
      validationFailures?: ValidationFailure[];
      resultName?: string;
    },
    submitting = false
  ): boolean {
    const updates: Partial<WebhookRouterFormState> = {};

    let ensureEmptyHeader = false;
    let toRemove: any[] = [];

    if (keys.hasOwnProperty('method')) {
      updates.method = { value: keys.method };

      const oldMethod = this.state.method.value.value;
      const newMethod = keys.method.value;

      if (oldMethod !== newMethod) {
        const existingContentTypeHeader = this.state.headers.find(
          (header: HeaderEntry) => header.value.name.toLowerCase() === 'content-type'
        );

        // whenever our method changes, update the default body
        updates.body = { value: getDefaultBody(newMethod) };

        // switching from a GET, add a content-type
        if (oldMethod === Methods.GET && newMethod !== Methods.GET) {
          if (!existingContentTypeHeader) {
            let uuid = createUUID();
            // if we have an empty header, use that one
            const lastHeader =
              this.state.headers.length > 0
                ? this.state.headers[this.state.headers.length - 1]
                : null;
            if (lastHeader && !lastHeader.value.name) {
              uuid = lastHeader.value.uuid;
            }
            keys.header = { uuid, name: 'Content-Type', value: 'application/json' };
          }
        } else if (oldMethod !== Methods.GET && newMethod === Methods.GET) {
          // remove content type if switching to a GET
          if (existingContentTypeHeader) {
            toRemove = [{ headers: [{ value: existingContentTypeHeader.value }] }];
          }
        }
      }
    }

    if (keys.hasOwnProperty('url')) {
      updates.url = validate(i18n.t('forms.url', 'URL'), keys.url, [
        shouldRequireIf(submitting),
        validateIf(ValidURL, keys.url.indexOf('@') === -1)
      ]);
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        shouldRequireIf(submitting),
        Alphanumeric,
        StartIsNonNumeric
      ]);
    }

    if (keys.hasOwnProperty('body')) {
      updates.body = { value: keys.body };
    }

    if (keys.hasOwnProperty('header')) {
      updates.headers = [{ value: keys.header, validationFailures: keys.validationFailures }];
      ensureEmptyHeader = true;
    }

    if (keys.hasOwnProperty('removeHeader')) {
      toRemove = [{ headers: [{ value: keys.removeHeader }] }];
      ensureEmptyHeader = true;
    }

    const updated = mergeForm(this.state, updates, toRemove);

    // update our form
    this.setState(updated, () => {
      // if we updated headers, check if we need a new one
      if (ensureEmptyHeader) {
        let needsHeader = true;
        for (const header of this.state.headers) {
          if (header.value.name.trim() === '') {
            needsHeader = false;
            break;
          }
        }

        if (needsHeader) {
          this.handleCreateHeader();
        }
      }
    });
    return updated.valid;
  }

  private handleUpdateResultName(value: string): boolean {
    return this.handleUpdate({ resultName: value });
  }

  private handleMethodUpdate(method: MethodOption): boolean {
    return this.handleUpdate({ method });
  }

  private handleUrlUpdate(url: string, name: string, submitting = false): boolean {
    return this.handleUpdate({ url }, submitting);
  }

  private handleHeaderRemoved(removeHeader: Header): boolean {
    return this.handleUpdate({ removeHeader });
  }

  private handleHeaderUpdated(header: Header, validationFailures: ValidationFailure[]): boolean {
    return this.handleUpdate({ header, validationFailures });
  }

  private handleCreateHeader(): boolean {
    return this.handleUpdate({
      header: {
        uuid: createUUID(),
        name: '',
        value: ''
      }
    });
  }

  private handleBodyUpdate(body: string): boolean {
    return this.handleUpdate({ body });
  }

  private handleSave(): void {
    // validate our url in case they haven't interacted
    const valid = this.handleUpdate(
      { url: this.state.url.value, resultName: this.state.resultName.value },
      true
    );

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const headerElements: JSX.Element[] = this.state.headers.map(
      (header: HeaderEntry, index: number, arr: HeaderEntry[]) => {
        return (
          <div key={`header_${header.value.uuid}`}>
            <HeaderElement
              entry={header}
              onRemove={this.handleHeaderRemoved}
              onChange={this.handleHeaderUpdated}
              index={index}
            />
          </div>
        );
      }
    );

    const tabs: Tab[] = [];
    tabs.push({
      name: i18n.t('forms.http_headers', 'HTTP Headers'),
      hasErrors: !!this.state.headers.find((header: HeaderEntry) => hasErrors(header)),
      body: (
        <>
          <p className={styles.info}>
            <Trans i18nKey="forms.webhook_header_summary">
              Add any additional headers below that you would like to send along with your request.
            </Trans>
          </p>
          {headerElements}
        </>
      ),
      checked: this.state.headers.length > 1
    });

    const method = this.state.method.value.value;
    const name = this.state.method.value.name + ' ' + i18n.t('body', 'Body');
    tabs.push({
      name,
      body: (
        <div key="post_body" className={styles.body_form}>
          <h4>{name}</h4>
          <p>
            <Trans
              i18nKey="forms.webhook_body_summary"
              values={{ method: this.state.method.value.name }}
            >
              Modify the body of the [[method]] request that will be sent to your webhook.
            </Trans>
          </p>
          <TextInputElement
            __className={styles.req_body}
            name={name}
            showLabel={false}
            entry={this.state.body}
            onChange={this.handleBodyUpdate}
            helpText={
              <Trans
                i18nKey="forms.webhook_body_summary"
                values={{ method: this.state.method.value.name }}
              >
                Modify the body of the [[method]] request that will be sent to your webhook.
              </Trans>
            }
            autocomplete={true}
            textarea={true}
          />
        </div>
      ),
      checked: this.state.body.value !== getDefaultBody(method)
    });

    return (
      <Dialog
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
        tabs={tabs}
      >
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <div style={{ display: 'flex', alignItems: 'flex-end' }}>
          <div className={styles.method}>
            <SelectElement
              key="method_select"
              name={i18n.t('forms.method', 'Method')}
              entry={this.state.method}
              onChange={this.handleMethodUpdate}
              options={METHOD_OPTIONS}
            />
          </div>
          <div className={styles.url}>
            <TextInputElement
              name={i18n.t('forms.url', 'URL')}
              placeholder={i18n.t('forms.enter_a_url', 'Enter a URL')}
              entry={this.state.url}
              onChange={this.handleUrlUpdate}
              autocomplete={true}
            />
          </div>
        </div>
        <div className={styles.instructions}>
          <p>
            <Trans i18nKey="forms.webhook_help">
              If your server responds with JSON, each property will be added to the Flow.
            </Trans>
          </p>
          <pre className={styles.code}>
            {'{ "product": "Solar Charging Kit", "stock level": 32 }'}
          </pre>
          <p>
            <Trans i18nKey="forms.webhook_example">
              This response would add <span className={styles.example}>@webhook.product</span> and{' '}
              <span className={styles.example}>@webhook["stock level"]</span> for use in the flow.
            </Trans>
          </p>
        </div>
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
