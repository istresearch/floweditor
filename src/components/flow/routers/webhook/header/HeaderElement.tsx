import { default as bindCallbacks } from 'auto-bind/react';
import { getAllErrors } from 'components/flow/actions/helpers';
import { HeaderEntry } from 'components/flow/routers/webhook/WebhookRouterForm';
import styles from 'components/flow/routers/webhook/WebhookRouterForm.module.scss';
import FormElement from 'components/form/FormElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import * as React from 'react';
import { StringEntry, ValidationFailure } from 'store/nodeEditor';
import { HeaderName, validate } from 'store/validators';
import i18n from 'config/i18n';

// TODO: move this into webhook router component
export interface Header {
  uuid: string;
  name: string;
  value: string;
}

export interface HeaderElementProps {
  entry: HeaderEntry;
  index: number;
  onRemove: (header: Header) => void;
  onChange: (header: Header, validationFailures: ValidationFailure[]) => void;
  empty?: boolean;
}

interface HeaderElementState {
  name: StringEntry;
  value: StringEntry;
}

export const headerContainerSpecId = 'header-container';
export const nameContainerSpecId = 'name-container';
export const valueConatainerSpecId = 'value-container';
export const removeIcoSpecId = 'remove-icon';

export const HEADER_NAME_ERROR = i18n.t(
  'errors.http_header_missing_name',
  'HTTP headers must have a name'
);
export const NAME_PLACEHOLDER = i18n.t('forms.webhook_header_name', 'Header Name');
export const VALUE_PLACEHOLDER = i18n.t('forms.value', 'Value');

export default class HeaderElement extends React.Component<HeaderElementProps, HeaderElementState> {
  constructor(props: HeaderElementProps) {
    super(props);

    const header = this.props.entry.value;
    const name = header.name || '';
    const value = header.value || '';

    this.state = {
      name: { value: name },
      value: { value }
    };

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  private getHeader(): Header {
    return {
      name: this.state.name.value,
      value: this.state.value.value,
      uuid: this.props.entry.value.uuid
    };
  }

  private handleChangeName(value: string): void {
    const name = validate(i18n.t('forms.header_name', 'Header name'), value, [HeaderName]);
    this.setState({ name: { value: name.value } }, () =>
      this.props.onChange(
        this.getHeader(),
        getAllErrors(this.state.value).concat(getAllErrors(name))
      )
    );
  }

  private handleChangeValue(value: string): void {
    this.setState({ value: { value } }, () => {
      const name = validate(i18n.t('forms.header_name', 'Header name'), this.state.name.value, [
        HeaderName
      ]);
      this.props.onChange(
        this.getHeader(),
        getAllErrors(this.state.value).concat(getAllErrors(name))
      );
    });
  }

  private handleRemove(): void {
    this.props.onRemove(this.getHeader());
  }

  private getRemoveIco(): JSX.Element {
    return (
      <div className={styles.remove_ico} onClick={this.handleRemove} data-spec={removeIcoSpecId}>
        <span className="fe-x" />
      </div>
    );
  }

  public render(): JSX.Element {
    const removeIco: JSX.Element = this.getRemoveIco();
    return (
      <FormElement name={i18n.t('forms.webhook_header', 'Header')} entry={this.props.entry}>
        <div className={styles.header} data-spec={headerContainerSpecId}>
          <div className={styles.header_name} data-spec={nameContainerSpecId}>
            <TextInputElement
              placeholder={NAME_PLACEHOLDER}
              name={NAME_PLACEHOLDER}
              onChange={this.handleChangeName}
              entry={this.state.name}
            />
          </div>
          <div className={styles.header_value} data-spec={valueConatainerSpecId}>
            <TextInputElement
              placeholder={VALUE_PLACEHOLDER}
              name={VALUE_PLACEHOLDER}
              onChange={this.handleChangeValue}
              entry={this.state.value}
              autocomplete={true}
            />
          </div>
          {removeIco}
        </div>
      </FormElement>
    );
  }
}
