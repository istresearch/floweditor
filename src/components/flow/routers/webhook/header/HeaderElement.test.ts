import HeaderElement, {
  headerContainerSpecId,
  HeaderElementProps,
  NAME_PLACEHOLDER,
  nameContainerSpecId,
  removeIcoSpecId,
  valueConatainerSpecId
} from 'components/flow/routers/webhook/header/HeaderElement';
import { HeaderEntry } from 'components/flow/routers/webhook/WebhookRouterForm';
import { composeComponentTestUtils, getSpecWrapper, setMock } from 'testUtils';
import { set, setFalse } from 'utils';

const headers: HeaderEntry[] = [
  {
    value: {
      uuid: '00c4498e-1c9e-4c26-aa7c-f81d13573129',
      name: 'Content-Type',
      value: 'application/json'
    }
  },
  {
    value: {
      uuid: '3e7d8366-22ff-452b-953e-3a9a88b5d72c ',
      name: 'Authorization',
      value:
        'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ'
    }
  },
  {
    value: { uuid: '236f2392-e576-4254-8c28-db175510c6a8', name: '', value: '' }
  }
];

const baseProps: HeaderElementProps = {
  entry: headers[0],
  index: 0,
  onRemove: jest.fn(),
  onChange: jest.fn()
};

const { setup, spyOn } = composeComponentTestUtils(HeaderElement, baseProps);

describe(HeaderElement.name, () => {
  describe('render', () => {
    it('should render self, children with base props', () => {
      const { wrapper, props, instance } = setup();
      const inputs = wrapper.find('TextInputElement');
      expect(getSpecWrapper(wrapper, headerContainerSpecId).hasClass('header')).toBeTruthy();

      expect(getSpecWrapper(wrapper, nameContainerSpecId).hasClass('header_name')).toBeTruthy();

      expect(inputs.at(0).props()).toEqual({
        placeholder: NAME_PLACEHOLDER,
        name: NAME_PLACEHOLDER,
        onChange: instance.handleChangeName,
        entry: { value: props.entry.value.name }
      });

      expect(getSpecWrapper(wrapper, valueConatainerSpecId).hasClass('header_value')).toBeTruthy();

      expect(inputs.at(1).props()).toMatchSnapshot();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render remove icon', () => {
      const { wrapper, instance } = setup(true, {
        index: set(1),
        empty: setFalse()
      });
      const removeIcon = getSpecWrapper(wrapper, removeIcoSpecId);

      expect(removeIcon.hasClass('remove_ico')).toBeTruthy();
      expect(wrapper.find('.fe-x').exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('instance methods', () => {
    describe('handleChangeName', () => {
      it('should update state, call onChange prop', () => {
        const setStateSpy = spyOn('setState');
        const {
          wrapper,
          props: { onChange: onChangeMock },
          instance
        } = setup(true, {
          onChange: setMock()
        });

        instance.handleChangeName(headers[0].value.name);
        wrapper.update();

        expect(setStateSpy).toHaveBeenCalledTimes(1);
        expect(setStateSpy).toMatchCallSnapshot('setState');
        expect(onChangeMock).toHaveBeenCalledTimes(1);
        expect(onChangeMock).toMatchCallSnapshot('onChange');

        expect(
          wrapper
            .find('TextInputElement')
            .at(0)
            .prop('entry')
        ).toEqual({ value: headers[0].value.name });
        expect(wrapper).toMatchSnapshot('text element');

        setStateSpy.mockRestore();
      });
    });

    describe('handleChangeValue', () => {
      it('should update state, call onChange prop', () => {
        const setStateSpy = spyOn('setState');
        const { wrapper, props, instance } = setup(true, {
          onChange: setMock()
        });

        instance.handleChangeValue(headers[0].value);
        wrapper.update();

        expect(setStateSpy).toHaveBeenCalledTimes(1);
        expect(setStateSpy).toMatchCallSnapshot();
        expect(props.onChange).toHaveBeenCalledTimes(1);
        expect(props.onChange).toMatchCallSnapshot('header change');

        expect(
          wrapper
            .find('TextInputElement')
            .at(1)
            .prop('entry')
        ).toEqual({ value: headers[0].value });
        expect(wrapper).toMatchSnapshot();

        setStateSpy.mockRestore();
      });
    });
  });
});
