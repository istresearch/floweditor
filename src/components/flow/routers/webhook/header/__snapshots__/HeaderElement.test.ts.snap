// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderElement instance methods handleChangeName should update state, call onChange prop: onChange 1`] = `
Array [
  Object {
    "name": "Content-Type",
    "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
    "value": "application/json",
  },
  Array [],
]
`;

exports[`HeaderElement instance methods handleChangeName should update state, call onChange prop: setState 1`] = `
Array [
  Object {
    "name": Object {
      "value": "Content-Type",
    },
  },
]
`;

exports[`HeaderElement instance methods handleChangeName should update state, call onChange prop: text element 1`] = `
<FormElement
  entry={
    Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
        "value": "application/json",
      },
    }
  }
  name="Header"
>
  <div
    className="header"
    data-spec="header-container"
  >
    <div
      className="header_name"
      data-spec="name-container"
    >
      <TextInputElement
        entry={
          Object {
            "value": "Content-Type",
          }
        }
        name="Header Name"
        onChange={[Function]}
        placeholder="Header Name"
      />
    </div>
    <div
      className="header_value"
      data-spec="value-container"
    >
      <TextInputElement
        autocomplete={true}
        entry={
          Object {
            "value": "application/json",
          }
        }
        name="Value"
        onChange={[Function]}
        placeholder="Value"
      />
    </div>
    <div
      className="remove_ico"
      data-spec="remove-icon"
      onClick={[Function]}
    >
      <span
        className="fe-x"
      />
    </div>
  </div>
</FormElement>
`;

exports[`HeaderElement instance methods handleChangeValue should update state, call onChange prop 1`] = `
Array [
  Object {
    "value": Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
        "value": "application/json",
      },
    },
  },
]
`;

exports[`HeaderElement instance methods handleChangeValue should update state, call onChange prop 2`] = `
<FormElement
  entry={
    Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
        "value": "application/json",
      },
    }
  }
  name="Header"
>
  <div
    className="header"
    data-spec="header-container"
  >
    <div
      className="header_name"
      data-spec="name-container"
    >
      <TextInputElement
        entry={
          Object {
            "value": "Content-Type",
          }
        }
        name="Header Name"
        onChange={[Function]}
        placeholder="Header Name"
      />
    </div>
    <div
      className="header_value"
      data-spec="value-container"
    >
      <TextInputElement
        autocomplete={true}
        entry={
          Object {
            "value": Object {
              "name": "Content-Type",
              "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
              "value": "application/json",
            },
          }
        }
        name="Value"
        onChange={[Function]}
        placeholder="Value"
      />
    </div>
    <div
      className="remove_ico"
      data-spec="remove-icon"
      onClick={[Function]}
    >
      <span
        className="fe-x"
      />
    </div>
  </div>
</FormElement>
`;

exports[`HeaderElement instance methods handleChangeValue should update state, call onChange prop: header change 1`] = `
Array [
  Object {
    "name": "Content-Type",
    "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
    "value": Object {
      "name": "Content-Type",
      "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
      "value": "application/json",
    },
  },
  Array [],
]
`;

exports[`HeaderElement render should render remove icon 1`] = `
<FormElement
  entry={
    Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
        "value": "application/json",
      },
    }
  }
  name="Header"
>
  <div
    className="header"
    data-spec="header-container"
  >
    <div
      className="header_name"
      data-spec="name-container"
    >
      <TextInputElement
        entry={
          Object {
            "value": "Content-Type",
          }
        }
        name="Header Name"
        onChange={[Function]}
        placeholder="Header Name"
      />
    </div>
    <div
      className="header_value"
      data-spec="value-container"
    >
      <TextInputElement
        autocomplete={true}
        entry={
          Object {
            "value": "application/json",
          }
        }
        name="Value"
        onChange={[Function]}
        placeholder="Value"
      />
    </div>
    <div
      className="remove_ico"
      data-spec="remove-icon"
      onClick={[Function]}
    >
      <span
        className="fe-x"
      />
    </div>
  </div>
</FormElement>
`;

exports[`HeaderElement render should render self, children with base props 1`] = `
Object {
  "autocomplete": true,
  "entry": Object {
    "value": "application/json",
  },
  "name": "Value",
  "onChange": [Function],
  "placeholder": "Value",
}
`;

exports[`HeaderElement render should render self, children with base props 2`] = `
<FormElement
  entry={
    Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "00c4498e-1c9e-4c26-aa7c-f81d13573129",
        "value": "application/json",
      },
    }
  }
  name="Header"
>
  <div
    className="header"
    data-spec="header-container"
  >
    <div
      className="header_name"
      data-spec="name-container"
    >
      <TextInputElement
        entry={
          Object {
            "value": "Content-Type",
          }
        }
        name="Header Name"
        onChange={[Function]}
        placeholder="Header Name"
      />
    </div>
    <div
      className="header_value"
      data-spec="value-container"
    >
      <TextInputElement
        autocomplete={true}
        entry={
          Object {
            "value": "application/json",
          }
        }
        name="Value"
        onChange={[Function]}
        placeholder="Value"
      />
    </div>
    <div
      className="remove_ico"
      data-spec="remove-icon"
      onClick={[Function]}
    >
      <span
        className="fe-x"
      />
    </div>
  </div>
</FormElement>
`;
